core/variables_dynamic.js

```diff
@@ -0,0 +1,127 @@
+ */
+Blockly.VariablesDynamic.promptType = function(promptText, defaultText, callback) {
+  Blockly.prompt(promptText, defaultText, function(newVarType) {
+    // Merge runs of whitespace.  Strip leading and trailing whitespace.
+    // Beyond this, all types are legal.
+    if (newVarType) {
+      newVarType = newVarType.replace(/[\s\xa0]+/g, ' ').replace(/^ | $/g, '');
+    }
+    callback(newVarType);
+  });
+};
+Blockly.VariablesDynamic.onCreateVariableButtonClick = function(button) {
+  Blockly.VariablesDynamic.promptType(Blockly.Msg.NEW_VARIABLE_TYPE_TITLE, '', function(type) {
+    if (type) {
+      Blockly.Variables.createVariable(button.getTargetWorkspace(), null, type);
+    }
+  });
+  // workspace.createVariable("abc", "string");
+  // workspace.createVariable("123", "number");
+  // workspace.createVariable("abcd", "string");
+};
+/**
+ * Construct the elements (blocks and button) required by the flyout for the
+ * variable category.
+ * @param {!Blockly.Workspace} workspace The workspace contianing variables.
+ * @return {!Array.<!Element>} Array of XML elements.
+ */
+Blockly.VariablesDynamic.flyoutCategory = function(workspace) {
+  var xmlList = [];
+  var button = goog.dom.createDom('button');
+  button.setAttribute('text', Blockly.Msg.NEW_VARIABLE);
+  button.setAttribute('callbackKey', 'CREATE_VARIABLE');
+
+  workspace.registerButtonCallback('CREATE_VARIABLE', Blockly.VariablesDynamic.onCreateVariableButtonClick);
+
+  xmlList.push(button);
+
+  var blockList = Blockly.VariablesDynamic.flyoutCategoryBlocks(workspace);
+  xmlList = xmlList.concat(blockList);
+  return xmlList;
+};
+
+/**
+ * Construct the blocks required by the flyout for the variable category.
+ * @param {!Blockly.Workspace} workspace The workspace contianing variables.
+ * @return {!Array.<!Element>} Array of XML block elements.
+ */
+Blockly.VariablesDynamic.flyoutCategoryBlocks = function(workspace) {
+  var variableModelList = workspace.getAllVariables();
+  variableModelList.sort(Blockly.VariableModel.compareByName);
+
+  var xmlList = [];
+  if (variableModelList.length > 0) {
+
+    var varTypes = workspace.getVariableTypes();
+    for (var i in varTypes) {
+      var varType = varTypes[i];
+      var variableModelListOfType = workspace.getVariablesOfType(varType);
+      var firstVariable = variableModelListOfType[0];
+      if (Blockly.Blocks['variables_set_dynamic']) {
+        var gap = i == varTypes.length - 1 ? 24 : 8;
+        var blockText = '<xml>' +
+                    '<block type="variables_set_dynamic" gap="' + gap + '">' +
+                    Blockly.Variables.generateVariableFieldXml_(firstVariable) +
+                    '</block>' +
+                    '</xml>';
+        var block = Blockly.Xml.textToDom(blockText).firstChild;
+        xmlList.push(block);
+      }
+    }
+    for (var i = 0, variable; variable = variableModelList[i]; i++) {
+      if (Blockly.Blocks['variables_get_dynamic']) {
```

Comment:
ditto on moving the check outside the loop. We should probably fix this in the other flyout too.